# Data Wrangling

In this project, we gathered our data from 7 different sources. So, we first need to inspect each dataset individually and make note of the overlaping variables.

In [1]:
#Importing all necessary packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas_profiling


We have `DataScientist.csv`, `DataAnalyst.csv`, `Data_Job_WA.csv`, `Data_Job_TX.csv`, `Data_Job_SF.csv`, `Data_Job_NY.csv` and `BusinessAnalyst.csv`. From inspecting the csv's, we can see that NA values were entered as '-1', so we are going to add the parameter 'na_values = -1'. Let's import them all to 7 different dataframes:

In [2]:
DataJobWA = pd.read_csv('Glassdoor job data/Data_Job_WA.csv', na_values=-1)
DataJobTX = pd.read_csv('Glassdoor job data/Data_Job_TX.csv', na_values=-1)
DataJobSF = pd.read_csv('Glassdoor job data/Data_Job_SF.csv', na_values=-1)
DataJobNY = pd.read_csv('Glassdoor job data/Data_Job_NY.csv', na_values=-1)

In [3]:
BusinessAnalyst = pd.read_csv('Glassdoor job data/BusinessAnalyst.csv', na_values=-1)

In [4]:
DataAnalyst = pd.read_csv('Glassdoor job data/DataAnalyst.csv', na_values=-1)

In [5]:
DataScientist = pd.read_csv('Glassdoor job data/DataScientist.csv', na_values=-1)

In [6]:
DataScientist.drop(columns = 'index',inplace=True)

In [7]:
BusinessAnalyst.drop(columns = 'index',inplace=True)

In [8]:
DataScientist

Unnamed: 0                                          Job Title  \
0              0                              Senior Data Scientist   
1              1                  Data Scientist, Product Analytics   
2              2                               Data Science Manager   
3              3                                       Data Analyst   
4              4                             Director, Data Science   
...          ...                                                ...   
3904        3904                                  AWS Data Engineer   
3905        3905                              Data Analyst â Junior   
3906        3906                   Security Analytics Data Engineer   
3907        3907                   Security Analytics Data Engineer   
3908        3908  Patient Safety Physician or Safety Scientist -...   

                   Salary Estimate  \
0     $111K-$181K (Glassdoor est.)   
1     $111K-$181K (Glassdoor est.)   
2     $111K-$181K (Glassdoor est.)   
3     $111K-$181K (Glassdoor est.)   
4     $111K-$181K (Glassdoor est.)   
...                            ...   
3904   $55K-$112K (Glassdoor est.)   
3905   $55K-$112K (Glassdoor est.)   
3906   $55K-$112K (Glassdoor est.)   
3907   $55K-$112K (Glassdoor est.)   
3908   $55K-$112K (Glassdoor est.)   

                                        Job Description  Rating  \
0     ABOUT HOPPER\n\nAt Hopper, we’re on a mission ...     3.5   
1     At Noom, we use scientifically proven methods ...     4.5   
2     Decode_M\n\nhttps://www.decode-m.com/\n\nData ...     NaN   
3     Sapphire Digital seeks a dynamic and driven mi...     3.4   
4     Director, Data Science - (200537)\nDescription...     3.4   
...                                                 ...     ...   
3904  About Us\n\nTachyon Technologies is a Digital ...     4.4   
3905  Job description\nInterpret data, analyze resul...     5.0   
3906  Job DescriptionThe Security Analytics Data Eng...     3.8   
3907  The Security Analytics Data Engineer will inte...     4.0   
3908  Help us transform patients' lives.\nAt UCB, we...     3.7   

                               Company Name       Location      Headquarters  \
0                               Hopper\n3.5   New York, NY  Montreal, Canada   
1                              Noom US\n4.5   New York, NY      New York, NY   
2                                  Decode_M   New York, NY      New York, NY   
3                     Sapphire Digital\n3.4  Lyndhurst, NJ     Lyndhurst, NJ   
4           United Entertainment Group\n3.4   New York, NY      New York, NY   
...                                     ...            ...               ...   
3904              Tachyon Technologies\n4.4     Dublin, OH        Irving, TX   
3905  Staffigo Technical Services, LLC\n5.0   Columbus, OH     Woodridge, IL   
3906                    PDS Tech, Inc.\n3.8     Dublin, OH        Irving, TX   
3907        Data Resource Technologies\n4.0     Dublin, OH         Omaha, NE   
3908                               UCB\n3.7     Slough, OH  Brussel, Belgium   

                         Size  Founded  Type of ownership  \
0       501 to 1000 employees   2007.0  Company - Private   
1      1001 to 5000 employees   2008.0  Company - Private   
2           1 to 50 employees      NaN            Unknown   
3        201 to 500 employees   2019.0  Company - Private   
4         51 to 200 employees   2007.0  Company - Private   
...                       ...      ...                ...   
3904     201 to 500 employees   2011.0  Company - Private   
3905      51 to 200 employees   2008.0  Company - Private   
3906  5001 to 10000 employees   1977.0  Company - Private   
3907        1 to 50 employees      NaN  Company - Private   
3908  5001 to 10000 employees      NaN   Company - Public   

                       Industry                     Sector  \
0               Travel Agencies           Travel & Tourism   
1     Health, Beauty, & Fitness          Consumer Services   
2          

In [9]:
DataAnalyst

Unnamed: 0                                          Job Title  \
0              0  Data Analyst, Center on Immigration and Justic...   
1              1                               Quality Data Analyst   
2              2  Senior Data Analyst, Insights & Analytics Team...   
3              3                                       Data Analyst   
4              4                             Reporting Data Analyst   
...          ...                                                ...   
2248        2248  RQS - IHHA - 201900004460 -1q Data Security An...   
2249        2249              Senior Data Analyst (Corporate Audit)   
2250        2250  Technical Business Analyst (SQL, Data analytic...   
2251        2251                Data Analyst 3, Customer Experience   
2252        2252                        Senior Quality Data Analyst   

                  Salary Estimate  \
0      $37K-$66K (Glassdoor est.)   
1      $37K-$66K (Glassdoor est.)   
2      $37K-$66K (Glassdoor est.)   
3      $37K-$66K (Glassdoor est.)   
4      $37K-$66K (Glassdoor est.)   
...                           ...   
2248  $78K-$104K (Glassdoor est.)   
2249  $78K-$104K (Glassdoor est.)   
2250  $78K-$104K (Glassdoor est.)   
2251  $78K-$104K (Glassdoor est.)   
2252  $78K-$104K (Glassdoor est.)   

                                        Job Description  Rating  \
0     Are you eager to roll up your sleeves and harn...     3.2   
1     Overview\n\nProvides analytical and technical ...     3.8   
2     We’re looking for a Senior Data Analyst who ha...     3.4   
3     Requisition NumberRR-0001939\nRemote:Yes\nWe c...     4.1   
4     ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...     3.9   
...                                                 ...     ...   
2248  Maintains systems to protect data from unautho...     2.5   
2249  Position:\nSenior Data Analyst (Corporate Audi...     2.9   
2250  Title: Technical Business Analyst (SQL, Data a...     NaN   
2251  Summary\n\nResponsible for working cross-funct...     3.1   
2252  You.\n\nYou bring your body, mind, heart and s...     3.4   

                                 Company Name        Location  \
0              Vera Institute of Justice\n3.2    New York, NY   
1     Visiting Nurse Service of New York\n3.8    New York, NY   
2                            Squarespace\n3.4    New York, NY   
3                               Celerity\n4.1    New York, NY   
4                                FanDuel\n3.9    New York, NY   
...                                       ...             ...   
2248                       Avacend, Inc.\n2.5      Denver, CO   
2249                   Arrow Electronics\n2.9  Centennial, CO   
2250                                 Spiceorb      Denver, CO   
2251         Contingent Network Services\n3.1  Centennial, CO   
2252                          SCL Health\n3.4  Broomfield, CO   

          Headquarters                    Size  Founded  \
0         New York, NY    201 to 500 employees   1961.0   
1         New York, NY        10000+ employees   1893.0   
2         New York, NY  1001 to 5000 employees   2003.0   
3           McLean, VA    201 to 500 employees   2002.0   
4         New York, NY   501 to 1000 employees   2009.0   
...                ...                     ...      ...   
2248    Alpharetta, GA     51 to 200 employees      NaN   
2249    Centennial, CO        10000+ employees   1935.0   
2250               NaN                     NaN      NaN   
2251  West Chester, OH    201 to 500 employees   1984.0   
2252    Broomfield, CO        10000+ employees   1864.0   

                   Type of ownership                                 Industry  \
0             Nonprofit Organization                        Social Assistance   
1             Nonprofit Organization         Health Care Services & Hospitals   
2                  Company - Private                                 Internet   
3     Subsidiary or Business Segment                              IT Services   
4       

In [10]:
BusinessAnalyst

Unnamed: 0  \
0                                                     0   
1                                                     1   
2                                                     2   
3                                                     3   
4                                                     4   
...                                                 ...   
4087                                  Help Desk Analyst   
4088              System Analyst (Data) - PL/SQl Server   
4089  Programmer Analyst- PeopleSoft ( Finance and S...   
4090                            Loss Mitigation Analyst   
4091                     Financial Analyst II - Baptist   

                                              Job Title  \
0      Business Analyst - Clinical & Logistics Platform   
1                                      Business Analyst   
2                                          Data Analyst   
3       Information Security Analyst, Incident Response   
4                         Analyst - FP&A Global Revenue   
...                                                 ...   
4087  Ref ID: 01120-0011488945Classification: Help D...   
4088  Job Title: System Analyst (Data) - PL/SQl Serv...   
4089  Job Opening Summary\nReports to the Systems Ap...   
4090  Job Description\nA knowledgeable job-seeker is...   
4091  Job Summary\n\nThis unique analyst position re...   

                  Salary Estimate  \
0     $56K-$102K (Glassdoor est.)   
1     $56K-$102K (Glassdoor est.)   
2     $56K-$102K (Glassdoor est.)   
3     $56K-$102K (Glassdoor est.)   
4     $56K-$102K (Glassdoor est.)   
...                           ...   
4087                          3.5   
4088                          3.9   
4089                          4.0   
4090                          4.4   
4091                          2.7   

                                        Job Description            Rating  \
0     Company Overview\n\n\nAt Memorial Sloan Ketter...               3.9   
1     We are seeking for an energetic and collaborat...               3.8   
2     For more than a decade, Asembia has been worki...               3.6   
3     Job Description Summary\nThe Information Secur...               3.6   
4     Magnite is the world's largest independent sel...               3.4   
...                                                 ...               ...   
4087                                   Robert Half\n3.5  Jacksonville, FL   
4088                                  Diverse Lynx\n3.9  Jacksonville, FL   
4089           Shands at the University of Florida\n4.0  Jacksonville, FL   
4090               Contemporary Staffing Solutions\n4.4  Jacksonville, FL   
4091           Baptist Medical Center Jacksonville\n2.7  Jacksonville, FL   

                       Company Name                 Location  \
0     Memorial Sloan-Kettering\n3.9             New York, NY   
1      Paine Schwartz Partners\n3.8             New York, NY   
2                      Asembia\n3.6         Florham Park, NJ   
3                           BD\n3.6       Franklin Lakes, NJ   
4              Rubicon Project\n3.4             New York, NY   
...                             ...                      ...   
4087                 Menlo Park, CA         10000+ employees   
4088                  Princeton, NJ    501 to 1000 employees   
4089                Gainesville, FL         10000+ employees   
4090               Mount Laurel, NJ   1001 to 5000 employees   
4091               Jacksonville, FL  5001 to 10000 employees   

            Headquarters                            Size  \
0           New York, NY                10000+ employees   
1           New York, NY               1 to 50 employees   
2       Florham Park, NJ           501 to 1000 employees   
3     Franklin Lakes, NJ                10000+ employees   
4        Los Angeles, CA            201 to 500 employees   
...                  ...                             ...   
4087                1948                Company - Public   
4088    

The dataframe `BusinessAnalyst`seems like at some point the variables got shifted. 

In [11]:
#Loop to find where the shift happened.
for i in range(len(BusinessAnalyst)):
    try:
        int(BusinessAnalyst.iloc[i,0])
    except:
        print(i)
        break

3692


It looks like it happened at row 3692. Let's check:

In [12]:
BusinessAnalyst.iloc[3691:,0]

3691                                                 3691
3692                Business Analyst - Loan IQ Specialist
3693                               Procure to Pay Analyst
3694                                      Payroll Analyst
3695                   Software Quality Assurance Analyst
                              ...                        
4087                                    Help Desk Analyst
4088                System Analyst (Data) - PL/SQl Server
4089    Programmer Analyst- PeopleSoft ( Finance and S...
4090                              Loss Mitigation Analyst
4091                       Financial Analyst II - Baptist
Name: Unnamed: 0, Length: 401, dtype: object

In [13]:
len(BusinessAnalyst) - 3692

400

In [14]:
BusinessAnalyst.drop(index= np.arange(3692,len(BusinessAnalyst),1), inplace =True)

In [15]:
print('--DataScientist: ', DataScientist.columns, '\n' '--DataAnalyst: ', DataAnalyst.columns, '\n' '--DataJobWA: ', DataJobWA.columns, '\n' '--DataJobTX: ', DataJobTX.columns, '\n' '--DataJobSF: ', 
      DataJobSF.columns, '\n' '--DataJobNY: ', DataJobNY.columns, '\n' '--BusinessAnalyst: ', BusinessAnalyst.columns)

--DataScientist:  Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'Job Description',
       'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply'],
      dtype='object') 
--DataAnalyst:  Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'Job Description',
       'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply'],
      dtype='object') 
--DataJobWA:  Index(['Job_title', 'Company', 'State', 'City', 'Min_Salary', 'Max_Salary',
       'Job_Desc', 'Industry', 'Rating', 'Date_Posted', 'Valid_until',
       'Job_Type'],
      dtype='object') 
--DataJobTX:  Index(['Job_title', 'Company', 'State', 'City', 'Min_Salary', 'Max_Salary',
       'Job_Desc', 'Industry', 'Rating', 'Date_Posted', 'Valid_until',
       'Job_Type'],
      dtype='object') 
--DataJobSF:

From this, we can see that `DataJobWA`, `DataJobTX`, `DataJobSF` and `DataJobNY` have the same variables, as for `DataScientist`, `BusinessAnalyst` and `DataAnalyst`:

In [16]:
#Let's write a function to compare all the column names of two given dataframes
def check_columns(df1, df2):
    columns1 = df1.columns
    columns2 = df2.columns
    matches = df1.columns == df2.columns
    if np.sum(matches) == len(matches):
        print("All columns are the same")
    else:
        print(columns1[~matches])

In [17]:
check_columns(DataScientist, DataAnalyst)

All columns are the same


In [18]:
check_columns(DataScientist, BusinessAnalyst)

All columns are the same


As we can see, we have two sets of dataframes that have the same variables. Now we need to compare both sets to see which variables are common in these sets:

In [19]:
for i in DataScientist.columns:
    if i not in DataJobNY.columns:
        print(i)

Unnamed: 0
Job Title
Salary Estimate
Job Description
Company Name
Location
Headquarters
Size
Founded
Type of ownership
Sector
Revenue
Competitors
Easy Apply


Here we can see we have a little issue: same variables are named differently in the two sets. 
We need to address this to be able to successfully merge all the dataframes together later.

In [20]:
DataScientist.head()

Unnamed: 0                          Job Title  \
0           0              Senior Data Scientist   
1           1  Data Scientist, Product Analytics   
2           2               Data Science Manager   
3           3                       Data Analyst   
4           4             Director, Data Science   

                Salary Estimate  \
0  $111K-$181K (Glassdoor est.)   
1  $111K-$181K (Glassdoor est.)   
2  $111K-$181K (Glassdoor est.)   
3  $111K-$181K (Glassdoor est.)   
4  $111K-$181K (Glassdoor est.)   

                                     Job Description  Rating  \
0  ABOUT HOPPER\n\nAt Hopper, we’re on a mission ...     3.5   
1  At Noom, we use scientifically proven methods ...     4.5   
2  Decode_M\n\nhttps://www.decode-m.com/\n\nData ...     NaN   
3  Sapphire Digital seeks a dynamic and driven mi...     3.4   
4  Director, Data Science - (200537)\nDescription...     3.4   

                      Company Name       Location      Headquarters  \
0                      Hopper\n3.5   New York, NY  Montreal, Canada   
1                     Noom US\n4.5   New York, NY      New York, NY   
2                         Decode_M   New York, NY      New York, NY   
3            Sapphire Digital\n3.4  Lyndhurst, NJ     Lyndhurst, NJ   
4  United Entertainment Group\n3.4   New York, NY      New York, NY   

                     Size  Founded  Type of ownership  \
0   501 to 1000 employees   2007.0  Company - Private   
1  1001 to 5000 employees   2008.0  Company - Private   
2       1 to 50 employees      NaN            Unknown   
3    201 to 500 employees   2019.0  Company - Private   
4     51 to 200 employees   2007.0  Company - Private   

                    Industry                  Sector  \
0            Travel Agencies        Travel & Tourism   
1  Health, Beauty, & Fitness       Consumer Services   
2                        NaN                     NaN   
3                   Internet  Information Technology   
4    Advertising & Marketing       Business Services   

                    Revenue               Competitors Easy Apply  
0  Unknown / Non-Applicable                       NaN        NaN  
1  Unknown / Non-Applicable                       NaN        NaN  
2  Unknown / Non-Applicable                       NaN       True  
3  Unknown / Non-Applicable      Zocdoc, Healthgrades        NaN  
4  Unknown / Non-Applicable  BBDO, Grey Group, Droga5        NaN

In [21]:
DataJobNY.head()

Job_title  \
0                      Chief Marketing Officer (CMO)   
1                                   Registered Nurse   
2                                   Dental Hygienist   
3                        Senior Salesforce Developer   
4  DEPUTY EXECUTIVE DIRECTOR, PROGRAM AND LEGAL A...   

                                 Company State           City  Min_Salary  \
0                   National Debt Relief    NY       New York         NaN   
1      Queens Boulevard Endoscopy Center    NY      Rego Park         NaN   
2                         Batista Dental    NJ  West New York         NaN   
3                   National Debt Relief    NY       New York     44587.0   
4  National Advocates for Pregnant Women    NY       New York    125410.0   

   Max_Salary                                           Job_Desc Industry  \
0         NaN  Who We're Looking For:\n\nThe Chief Marketing ...  Finance   
1         NaN  Queens Boulevard Endoscopy Center, an endoscop...      NaN   
2         NaN  Part-time or Full-timedental hygienist positio...      NaN   
3     82162.0  Principle Duties & Responsibilities:\n\nAnalyz...  Finance   
4    212901.0  For FULL Job Announcement, visit our website: ...      NaN   

   Rating Date_Posted Valid_until   Job_Type  
0     4.0  2020-05-08  2020-06-07  FULL_TIME  
1     3.0  2020-04-25  2020-06-07  FULL_TIME  
2     NaN  2020-05-02  2020-06-07  PART_TIME  
3     4.0  2020-05-08  2020-06-07  FULL_TIME  
4     NaN  2020-04-28  2020-06-07  FULL_TIME

Here we can see we have a lot to do. The columns from `DataJobNY` seem to have a better naming convention and be organized a little better for our purpose. That being said, some variables need to be renamed like `Job Title` and `Job_tile` to be the same. `Location` needs to be broken up into `State` and `City`. `Salary Estimate` should be split into `Min_Salary` and `Max_Salary`.


In [22]:
DataScientist.rename(columns={'Job Title':'Job_title', 'Job Description': 'Job_Desc', 'Company Name' : 'Company'}, inplace=True)
BusinessAnalyst.rename(columns={'Job Title':'Job_title', 'Job Description': 'Job_Desc', 'Company Name' : 'Company'}, inplace=True)
DataAnalyst.rename(columns={'Job Title':'Job_title', 'Job Description': 'Job_Desc', 'Company Name' : 'Company'}, inplace=True)

In [23]:
for i in DataScientist.columns:
    if i not in DataJobNY.columns:
        print(i)

Unnamed: 0
Salary Estimate
Location
Headquarters
Size
Founded
Type of ownership
Sector
Revenue
Competitors
Easy Apply


In [24]:
for i in DataJobNY.columns:
    if i not in DataScientist.columns:
        print(i)

State
City
Min_Salary
Max_Salary
Date_Posted
Valid_until
Job_Type


Here we can see all the variables that are missing from each dataset. We already stablished that some of them need to be broken up, and after that, we need to decide what to do with the variables that were not present on both sets of dataframes from the beginning.

Let's start with `Location`:

In [25]:
def location(df):
    df['In_Headquarters'] = df.apply(lambda x: 1 if x.Location == x.Headquarters else 0, axis=1)
    df.Location = df.Location.str.split(',')
    df['City'] = [i[0] for i in df.Location]
    df['State'] = [i[1].strip() for i in df.Location]
    #Reverting the split so we don't encounter issues with df operations
    df['Location'] = df['Location'].str.join(',')
    #df.drop(columns = 'Location', inplace=True)
    return df

Let's check for missing values first:

In [26]:
DataAnalyst[DataAnalyst['Location'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply]
Index: []

In [27]:
DataScientist[DataScientist['Location'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply]
Index: []

In [28]:
BusinessAnalyst[BusinessAnalyst['Location'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply]
Index: []

In [29]:
location(DataAnalyst);

In [30]:
location(DataScientist);

In [31]:
location(BusinessAnalyst);

Now let's break up `Salary Estimate`:

In [32]:
#Function to split 'Salary Estimate into 'Min_Salary' and 'Max_Salary'. Also removes '$' and 'K', turns the variable type into integer and multiplies by 1000
def salary(df):
    df['Salary Estimate'] = df['Salary Estimate'].str.split('-')
    df['Min_Salary'] = [i[0].strip('$') for i in df['Salary Estimate']]
    df['Max_Salary'] = [i[1].strip('$') for i in df['Salary Estimate']]
    df['Max_Salary'] = df['Max_Salary'].str.split('(')
    df['Employer_Estimate'] = [0 if i[1].find('mploy') == -1 else 1 for i in df['Max_Salary']]
    df['Max_Salary'] = [i[0].strip() for i in df['Max_Salary']]
    #Reverting the split so we don't encounter issues with df operations
    df['Salary Estimate'] = df['Salary Estimate'].str.join('-')
    #df.drop(columns = 'Salary Estimate', inplace=True)
    df['Min_Salary'] = df.Min_Salary.str.strip('K').astype(int)*1000
    df['Max_Salary'] = df.Max_Salary.str.strip('K').astype(int)*1000
    return df

Let's see if any of the entries are missing the salary estimate:

In [33]:
DataAnalyst[DataAnalyst['Salary Estimate'].isnull()]

Unnamed: 0                                  Job_title Salary Estimate  \
2149        2149  Configuration/Data Management Analyst III             NaN   

                                               Job_Desc  Rating  \
2149  Job Description\nPosition Title: Configuration...     4.4   

              Company  Location  Headquarters                 Size  Founded  \
2149  Protingent\n4.4  Kent, WA  Bellevue, WA  51 to 200 employees   2001.0   

      Type of ownership                Industry             Sector  \
2149  Company - Private  Staffing & Outsourcing  Business Services   

                       Revenue Competitors Easy Apply  In_Headquarters  City  \
2149  $25 to $50 million (USD)         NaN        NaN                0  Kent   

     State  
2149    WA

In [34]:
DataScientist[DataScientist['Salary Estimate'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply, In_Headquarters, City, State]
Index: []

In [35]:
BusinessAnalyst[BusinessAnalyst['Salary Estimate'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply, In_Headquarters, City, State]
Index: []

Only `DataAnalyst` has one missing value... let's drop it:

In [36]:
DataAnalyst.drop(index=2149, inplace=True)

Let's check for hourly rates rather than salary:

In [37]:
DataAnalyst[DataAnalyst['Salary Estimate'].str.contains('Per', regex=False)].head()

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply, In_Headquarters, City, State]
Index: []

In [38]:
BusinessAnalyst[BusinessAnalyst['Salary Estimate'].str.contains('Per', regex=False)].head()

Empty DataFrame
Columns: [Unnamed: 0, Job_title, Salary Estimate, Job_Desc, Rating, Company, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors, Easy Apply, In_Headquarters, City, State]
Index: []

In [39]:
DataScientist[DataScientist['Salary Estimate'].str.contains('Per', regex=False)].head()

Unnamed: 0                                          Job_title  \
685         685            Clinical Laboratory Scientist, Per Diem   
686         686                                   Sr. Data Analyst   
687         687       Lead Data Science Instructor, Data Scientist   
688         688  Clinical Laboratory Scientist (Per Diem, Varia...   
689         689                                      Data Engineer   

                      Salary Estimate  \
685  $34-$53 Per Hour(Glassdoor est.)   
686  $34-$53 Per Hour(Glassdoor est.)   
687  $34-$53 Per Hour(Glassdoor est.)   
688  $34-$53 Per Hour(Glassdoor est.)   
689  $34-$53 Per Hour(Glassdoor est.)   

                                              Job_Desc  Rating  \
685  As the Clinical Laboratory Scientist, you will...     3.8   
686  As a Sr. Data Analyst, you will work closely w...     2.8   
687  Job Description\nGalvanize is a dynamic learni...     NaN   
688  Monterey Park Hospital, located in the San Gab...     3.0   
689  Responsibilities:* Responsible for building an...     NaN   

                                               Company            Location  \
685                       UC Davis Medical Center\n3.8     Los Angeles, CA   
686  SAG-AFTRA Health Plan and SAG-Producers Pensio...         Burbank, CA   
687                                     Galvanize Inc.     Los Angeles, CA   
688                               AHMC Healthcare\n3.0   Monterey Park, CA   
689                                  BlackLine Systems  Woodland Hills, CA   

        Headquarters                     Size  Founded  \
685   Sacramento, CA         10000+ employees      NaN   
686      Burbank, CA                  Unknown      NaN   
687   Harrisburg, NC      51 to 200 employees      NaN   
688     Alhambra, CA  5001 to 10000 employees   1998.0   
689  Markham, Canada        1 to 50 employees      NaN   

          Type of ownership                          Industry  \
685  Nonprofit Organization  Health Care Services & Hospitals   
686       Company - Private                               NaN   
687       Company - Private                      Construction   
688       Company - Private  Health Care Services & Hospitals   
689                 Unknown                               NaN   

                                 Sector                   Revenue Competitors  \
685                         Health Care    $2 to $5 billion (USD)         NaN   
686                                 NaN  Unknown / Non-Applicable         NaN   
687  Construction, Repair & Maintenance  Unknown / Non-Applicable         NaN   
688                         Health Care    $1 to $2 billion (USD)         NaN   
689                                 NaN  Unknown / Non-Applicable         NaN   

    Easy Apply  In_Headquarters            City State  
685        NaN                0     Los Angeles    CA  
686        NaN                1         Burbank    CA  
687        NaN                0     Los Angeles    CA  
688        NaN                0   Monterey Park    CA  
689        NaN                0  Woodland Hills    CA

DataScientist seems to have some entries with hourly rate, let's see how significant they are and drop them if they are not:

In [40]:
DataScientist['Salary Estimate'].str.contains('Per', regex=False).sum()

21

Only 21 entries are in hourly rate, so we can drop them:

In [41]:
DataScientist = DataScientist[~DataScientist['Salary Estimate'].str.contains('Per', regex=False).to_numpy(dtype=bool)].copy()

In [42]:
salary(BusinessAnalyst);

In [43]:
salary(DataAnalyst);

In [44]:
salary(DataScientist);

In [45]:
#Dropping 'Unnamed: 0'
DataScientist.drop(columns= 'Unnamed: 0', inplace=True)
DataAnalyst.drop(columns= 'Unnamed: 0', inplace=True)
BusinessAnalyst.drop(columns= 'Unnamed: 0', inplace=True)

Let's check how we are doing:

In [46]:
for i in DataScientist.columns:
    if i not in DataJobNY.columns:
        print(i)

Salary Estimate
Location
Headquarters
Size
Founded
Type of ownership
Sector
Revenue
Competitors
Easy Apply
In_Headquarters
Employer_Estimate


In [47]:
for i in DataJobNY.columns:
    if i not in DataScientist.columns:
        print(i)

Date_Posted
Valid_until
Job_Type


Now let's get some insights:

In [48]:
DataScientist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3888 entries, 0 to 3908
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          3888 non-null   object 
 1   Salary Estimate    3888 non-null   object 
 2   Job_Desc           3888 non-null   object 
 3   Rating             3483 non-null   float64
 4   Company            3888 non-null   object 
 5   Location           3888 non-null   object 
 6   Headquarters       3648 non-null   object 
 7   Size               3659 non-null   object 
 8   Founded            2918 non-null   float64
 9   Type of ownership  3659 non-null   object 
 10  Industry           3345 non-null   object 
 11  Sector             3345 non-null   object 
 12  Revenue            3659 non-null   object 
 13  Competitors        1145 non-null   object 
 14  Easy Apply         163 non-null    object 
 15  In_Headquarters    3888 non-null   int64  
 16  City               3888 

In [49]:
100*DataScientist.Competitors.isnull().sum()/len(DataScientist)

70.55041152263375

DataScientist is missing 70% of the entries in `Competitors` and almost all of them in `Easy Apply`. Since we will end up concatinating all these dataframes, let's check them all before deciding if we impute or drop them:

In [50]:
DataAnalyst.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2252 entries, 0 to 2252
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          2252 non-null   object 
 1   Salary Estimate    2252 non-null   object 
 2   Job_Desc           2252 non-null   object 
 3   Rating             1980 non-null   float64
 4   Company            2251 non-null   object 
 5   Location           2252 non-null   object 
 6   Headquarters       2080 non-null   object 
 7   Size               2089 non-null   object 
 8   Founded            1592 non-null   float64
 9   Type of ownership  2089 non-null   object 
 10  Industry           1899 non-null   object 
 11  Sector             1899 non-null   object 
 12  Revenue            2089 non-null   object 
 13  Competitors        521 non-null    object 
 14  Easy Apply         80 non-null     object 
 15  In_Headquarters    2252 non-null   int64  
 16  City               2252 

In [51]:
BusinessAnalyst.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3692 entries, 0 to 3691
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Job_title          3692 non-null   object
 1   Salary Estimate    3692 non-null   object
 2   Job_Desc           3692 non-null   object
 3   Rating             3374 non-null   object
 4   Company            3692 non-null   object
 5   Location           3692 non-null   object
 6   Headquarters       3495 non-null   object
 7   Size               3505 non-null   object
 8   Founded            2745 non-null   object
 9   Type of ownership  3505 non-null   object
 10  Industry           3227 non-null   object
 11  Sector             3227 non-null   object
 12  Revenue            3505 non-null   object
 13  Competitors        1016 non-null   object
 14  Easy Apply         133 non-null    object
 15  In_Headquarters    3692 non-null   int64 
 16  City               3692 non-null   object


In [52]:
DataJobWA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Job_title    892 non-null    object 
 1   Company      892 non-null    object 
 2   State        890 non-null    object 
 3   City         892 non-null    object 
 4   Min_Salary   674 non-null    float64
 5   Max_Salary   674 non-null    float64
 6   Job_Desc     892 non-null    object 
 7   Industry     761 non-null    object 
 8   Rating       794 non-null    float64
 9   Date_Posted  892 non-null    object 
 10  Valid_until  892 non-null    object 
 11  Job_Type     892 non-null    object 
dtypes: float64(3), object(9)
memory usage: 83.8+ KB


In [53]:
DataJobNY.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Job_title    900 non-null    object 
 1   Company      900 non-null    object 
 2   State        900 non-null    object 
 3   City         900 non-null    object 
 4   Min_Salary   455 non-null    float64
 5   Max_Salary   455 non-null    float64
 6   Job_Desc     900 non-null    object 
 7   Industry     600 non-null    object 
 8   Rating       660 non-null    float64
 9   Date_Posted  900 non-null    object 
 10  Valid_until  900 non-null    object 
 11  Job_Type     900 non-null    object 
dtypes: float64(3), object(9)
memory usage: 84.5+ KB


In [54]:
DataJobTX.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 643 entries, 0 to 642
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Job_title    643 non-null    object 
 1   Company      643 non-null    object 
 2   State        643 non-null    object 
 3   City         637 non-null    object 
 4   Min_Salary   470 non-null    float64
 5   Max_Salary   470 non-null    float64
 6   Job_Desc     643 non-null    object 
 7   Industry     572 non-null    object 
 8   Rating       587 non-null    float64
 9   Date_Posted  643 non-null    object 
 10  Valid_until  643 non-null    object 
 11  Job_Type     643 non-null    object 
dtypes: float64(3), object(9)
memory usage: 60.4+ KB


In [55]:
DataJobSF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Job_title    889 non-null    object 
 1   Company      889 non-null    object 
 2   State        889 non-null    object 
 3   City         889 non-null    object 
 4   Min_Salary   633 non-null    float64
 5   Max_Salary   633 non-null    float64
 6   Job_Desc     889 non-null    object 
 7   Industry     767 non-null    object 
 8   Rating       808 non-null    float64
 9   Date_Posted  889 non-null    object 
 10  Valid_until  889 non-null    object 
 11  Job_Type     889 non-null    object 
dtypes: float64(3), object(9)
memory usage: 83.5+ KB


Everything else seems okay at first glance... `Min_Salary` and `Max_Salary` are missing about 50% of the entries in DataJobNY, but we'll worry about that later. For now, we drop `Competitors` and `Easy Apply` from all the dataframes that have them:

In [56]:
DataScientist.drop(columns=['Competitors','Easy Apply'], inplace=True)
DataAnalyst.drop(columns=['Competitors','Easy Apply'], inplace=True)
BusinessAnalyst.drop(columns=['Competitors','Easy Apply'], inplace=True)

We should be in a good place to join all the dataframes:

In [57]:
df= pd.DataFrame()
df = df.append(DataScientist)
df = df.append(DataAnalyst)
df = df.append(BusinessAnalyst)
df = df.append(DataJobNY)
df = df.append(DataJobTX)
df = df.append(DataJobWA)
df = df.append(DataJobSF)

Reset index so there are no repeated indices:

In [58]:
df.reset_index(inplace=True)

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13156 entries, 0 to 13155
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   index              13156 non-null  int64  
 1   Job_title          13156 non-null  object 
 2   Salary Estimate    9832 non-null   object 
 3   Job_Desc           13156 non-null  object 
 4   Rating             11686 non-null  object 
 5   Company            13155 non-null  object 
 6   Location           9832 non-null   object 
 7   Headquarters       9223 non-null   object 
 8   Size               9253 non-null   object 
 9   Founded            7255 non-null   object 
 10  Type of ownership  9253 non-null   object 
 11  Industry           11171 non-null  object 
 12  Sector             8471 non-null   object 
 13  Revenue            9253 non-null   object 
 14  In_Headquarters    9832 non-null   float64
 15  City               13150 non-null  object 
 16  State              131

Here we can easily determine `Date_Posted`, `Valid_until` and ` Job_Type` need to be dropped as they are missing well over 75% of the entries:

In [60]:
df.drop(columns=['Date_Posted','Valid_until', 'Job_Type','index'], inplace=True)

Now, let's explore our data:

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13156 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          13156 non-null  object 
 1   Salary Estimate    9832 non-null   object 
 2   Job_Desc           13156 non-null  object 
 3   Rating             11686 non-null  object 
 4   Company            13155 non-null  object 
 5   Location           9832 non-null   object 
 6   Headquarters       9223 non-null   object 
 7   Size               9253 non-null   object 
 8   Founded            7255 non-null   object 
 9   Type of ownership  9253 non-null   object 
 10  Industry           11171 non-null  object 
 11  Sector             8471 non-null   object 
 12  Revenue            9253 non-null   object 
 13  In_Headquarters    9832 non-null   float64
 14  City               13150 non-null  object 
 15  State              13154 non-null  object 
 16  Min_Salary         120

In [62]:
df.Rating = df.Rating.astype(float)

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13156 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          13156 non-null  object 
 1   Salary Estimate    9832 non-null   object 
 2   Job_Desc           13156 non-null  object 
 3   Rating             11686 non-null  float64
 4   Company            13155 non-null  object 
 5   Location           9832 non-null   object 
 6   Headquarters       9223 non-null   object 
 7   Size               9253 non-null   object 
 8   Founded            7255 non-null   object 
 9   Type of ownership  9253 non-null   object 
 10  Industry           11171 non-null  object 
 11  Sector             8471 non-null   object 
 12  Revenue            9253 non-null   object 
 13  In_Headquarters    9832 non-null   float64
 14  City               13150 non-null  object 
 15  State              13154 non-null  object 
 16  Min_Salary         120

Let's do some cleanup: Some company names have the rating after '\n' and some entries have `Revenue`, `Size` and `Type of ownsership` as 'Unknown'  

In [64]:
df.Company = df.Company.str.rstrip("\n1234567890.")

In [65]:
df.loc[df.Revenue.str.contains('Unk', regex=False, na=True), 'Revenue'] = np.NaN

In [66]:
df.loc[df.Size.str.contains('Unk', regex=False, na=True), 'Size'] = np.NaN

In [67]:
df.loc[df['Type of ownership'].str.contains('Unk', regex=False, na=True), 'Type of ownership'] = np.NaN

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13156 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          13156 non-null  object 
 1   Salary Estimate    9832 non-null   object 
 2   Job_Desc           13156 non-null  object 
 3   Rating             11686 non-null  float64
 4   Company            13155 non-null  object 
 5   Location           9832 non-null   object 
 6   Headquarters       9223 non-null   object 
 7   Size               9049 non-null   object 
 8   Founded            7255 non-null   object 
 9   Type of ownership  9157 non-null   object 
 10  Industry           11171 non-null  object 
 11  Sector             8471 non-null   object 
 12  Revenue            6620 non-null   object 
 13  In_Headquarters    9832 non-null   float64
 14  City               13150 non-null  object 
 15  State              13154 non-null  object 
 16  Min_Salary         120

### Duplicates
Now, we check for duplicates:

In [69]:
df[df.duplicated()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 875 entries, 5165 to 11760
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          875 non-null    object 
 1   Salary Estimate    2 non-null      object 
 2   Job_Desc           875 non-null    object 
 3   Rating             643 non-null    float64
 4   Company            875 non-null    object 
 5   Location           2 non-null      object 
 6   Headquarters       2 non-null      object 
 7   Size               2 non-null      object 
 8   Founded            1 non-null      object 
 9   Type of ownership  2 non-null      object 
 10  Industry           584 non-null    object 
 11  Sector             1 non-null      object 
 12  Revenue            1 non-null      object 
 13  In_Headquarters    2 non-null      float64
 14  City               875 non-null    object 
 15  State              875 non-null    object 
 16  Min_Salary         44

There seems to be 875 duplicated entries, so we drop them:

In [70]:
df.drop_duplicates(inplace=True)

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12281 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          12281 non-null  object 
 1   Salary Estimate    9830 non-null   object 
 2   Job_Desc           12281 non-null  object 
 3   Rating             11043 non-null  float64
 4   Company            12280 non-null  object 
 5   Location           9830 non-null   object 
 6   Headquarters       9221 non-null   object 
 7   Size               9047 non-null   object 
 8   Founded            7254 non-null   object 
 9   Type of ownership  9155 non-null   object 
 10  Industry           10587 non-null  object 
 11  Sector             8470 non-null   object 
 12  Revenue            6619 non-null   object 
 13  In_Headquarters    9830 non-null   float64
 14  City               12275 non-null  object 
 15  State              12279 non-null  object 
 16  Min_Salary         116

Let's check some of the response variables with missing values:

In [72]:
df[df.Min_Salary.isnull()].head()

Job_title Salary Estimate  \
9832           Chief Marketing Officer (CMO)             NaN   
9833                        Registered Nurse             NaN   
9834                        Dental Hygienist             NaN   
9841                                 Plumber             NaN   
9844  Principal Incident Response Consultant             NaN   

                                               Job_Desc  Rating  \
9832  Who We're Looking For:\n\nThe Chief Marketing ...     4.0   
9833  Queens Boulevard Endoscopy Center, an endoscop...     3.0   
9834  Part-time or Full-timedental hygienist positio...     NaN   
9841  Established plumbing company looking for a Plu...     NaN   
9844  JOB DESCRIPTIONCompany background\n\nThe Cryps...     4.8   

                                Company Location Headquarters Size Founded  \
9832               National Debt Relief      NaN          NaN  NaN     NaN   
9833  Queens Boulevard Endoscopy Center      NaN          NaN  NaN     NaN   
9834                     Batista Dental      NaN          NaN  NaN     NaN   
9841              Plumbing for life Inc      NaN          NaN  NaN     NaN   
9844                      Crypsis Group      NaN          NaN  NaN     NaN   

     Type of ownership                Industry Sector Revenue  \
9832               NaN                 Finance    NaN     NaN   
9833               NaN                     NaN    NaN     NaN   
9834               NaN                     NaN    NaN     NaN   
9841               NaN                     NaN    NaN     NaN   
9844               NaN  Information Technology    NaN     NaN   

      In_Headquarters           City State  Min_Salary  Max_Salary  \
9832              NaN       New York    NY         NaN         NaN   
9833              NaN      Rego Park    NY         NaN         NaN   
9834              NaN  West New York    NJ         NaN         NaN   
9841              NaN       New York    NY         NaN         NaN   
9844              NaN       New York    NY         NaN         NaN   

      Employer_Estimate  
9832                NaN  
9833                NaN  
9834                NaN  
9841                NaN  
9844                NaN

Interesting... we seem to have job postings that are not related to our study population (Data Science), so let's filter out those positions we don't care about:

In [73]:
df.apply(lambda x: True if 'data' in x.Job_Desc.lower() or 'analyst' in x.Job_Desc.lower() or 'data scientist' in x.Job_Desc.lower() 
            or 'learning' in x.Job_Desc.lower() or'data science' in x.Job_Desc.lower() or 'statisti' in x.Job_Desc.lower() or 'sql' in x.Job_Desc.lower() or 'python' in x.Job_Desc.lower()
            or 'data' in x.Job_title.lower() or 'analyst' in x.Job_title.lower() or 'data scientist' in x.Job_title.lower() 
            or 'learning' in x.Job_title.lower() or'data science' in x.Job_title.lower() else False, axis =1).value_counts()

True     12066
False      215
dtype: int64

We have 823 entries that don't include any of the keywords we specified as relevant to our study population. Let's randomly check the job description of a couple of them and thendrop them and end up with 12333 entries:

In [74]:
df.loc[~df.apply(lambda x: True if 'data' in x.Job_Desc.lower() or 'analyst' in x.Job_Desc.lower() or 'data scientist' in x.Job_Desc.lower() 
            or 'learning' in x.Job_Desc.lower() or'data science' in x.Job_Desc.lower() or 'statisti' in x.Job_Desc.lower() or 'sql' in x.Job_Desc.lower() or 'python' in x.Job_Desc.lower()
            or 'data' in x.Job_title.lower() or 'analyst' in x.Job_title.lower() or 'data scientist' in x.Job_title.lower() 
            or 'learning' in x.Job_title.lower() or'data science' in x.Job_title.lower() else False, axis =1),'Job_title'].sort_index()

601                           Principal Research Scientist
639      Senior Thermal Modeling Engineer(Relocation to...
643                              Senior Research Scientist
1047                Valuation & Business Analytics Manager
1126                              Web Analytics Consultant
                               ...                        
13123                  Bioinformatics Scientist / Engineer
13125                          Scientist II - Gene Editing
13134    Sr. Research Scientist II, Formulation Process...
13135    Sr Scientist II/Principal Research Scientist I...
13149             Investigator - Senior Research Scientist
Name: Job_title, Length: 215, dtype: object

In [75]:
#df.iloc[10728,3]
#df.iloc[9953,3]
#df.iloc[9840,3]
df.iloc[10272,3]

3.9

None of the randomly sampled job descriptions were relevant to our study.

In [76]:
df = df[df.apply(lambda x: True if 'data' in x.Job_Desc.lower() or 'analyst' in x.Job_Desc.lower() or 'data scientist' in x.Job_Desc.lower() 
            or 'learning' in x.Job_Desc.lower() or'data science' in x.Job_Desc.lower() or 'statisti' in x.Job_Desc.lower() or 'sql' in x.Job_Desc.lower() or 'python' in x.Job_Desc.lower()
            or 'data' in x.Job_title.lower() or 'analyst' in x.Job_title.lower() or 'data scientist' in x.Job_title.lower() 
            or 'learning' in x.Job_title.lower() or'data science' in x.Job_title.lower() else False, axis =1)].copy()

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12066 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          12066 non-null  object 
 1   Salary Estimate    9769 non-null   object 
 2   Job_Desc           12066 non-null  object 
 3   Rating             10846 non-null  float64
 4   Company            12065 non-null  object 
 5   Location           9769 non-null   object 
 6   Headquarters       9165 non-null   object 
 7   Size               8994 non-null   object 
 8   Founded            7206 non-null   object 
 9   Type of ownership  9100 non-null   object 
 10  Industry           10392 non-null  object 
 11  Sector             8417 non-null   object 
 12  Revenue            6571 non-null   object 
 13  In_Headquarters    9769 non-null   float64
 14  City               12061 non-null  object 
 15  State              12064 non-null  object 
 16  Min_Salary         114

We have almost 500 missing values in our response variables. Here we need to decide whether to impute them or drop them. For this case, we will drop them, since in order to impute them we will need to get the mean for `Min_Salary` and `Max_Salary` for each `Job_title`, at least, or we would be sabotaging our model. 

In [78]:
df.dropna(subset = ['Min_Salary'], inplace = True)

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11444 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job_title          11444 non-null  object 
 1   Salary Estimate    9769 non-null   object 
 2   Job_Desc           11444 non-null  object 
 3   Rating             10395 non-null  float64
 4   Company            11443 non-null  object 
 5   Location           9769 non-null   object 
 6   Headquarters       9165 non-null   object 
 7   Size               8994 non-null   object 
 8   Founded            7206 non-null   object 
 9   Type of ownership  9100 non-null   object 
 10  Industry           9981 non-null   object 
 11  Sector             8417 non-null   object 
 12  Revenue            6571 non-null   object 
 13  In_Headquarters    9769 non-null   float64
 14  City               11444 non-null  object 
 15  State              11444 non-null  object 
 16  Min_Salary         114

Let's look at out categorical variables with missing values: 

In [80]:
df[['Type of ownership','Size','Industry','Sector','Revenue']]

Type of ownership                    Size                   Industry  \
0      Company - Private   501 to 1000 employees            Travel Agencies   
1      Company - Private  1001 to 5000 employees  Health, Beauty, & Fitness   
2                    NaN       1 to 50 employees                        NaN   
3      Company - Private    201 to 500 employees                   Internet   
4      Company - Private     51 to 200 employees    Advertising & Marketing   
...                  ...                     ...                        ...   
13150                NaN                     NaN                  Education   
13151                NaN                     NaN                        NaN   
13152                NaN                     NaN                        NaN   
13153                NaN                     NaN     Information Technology   
13155                NaN                     NaN     Information Technology   

                       Sector Revenue  
0            Travel & Tourism     NaN  
1           Consumer Services     NaN  
2                         NaN     NaN  
3      Information Technology     NaN  
4           Business Services     NaN  
...                       ...     ...  
13150                     NaN     NaN  
13151                     NaN     NaN  
13152                     NaN     NaN  
13153                     NaN     NaN  
13155                     NaN     NaN  

[11444 rows x 5 columns]

In [81]:
df.loc[~df.Sector.str.strip().duplicated(),'Sector'].count(), pd.unique(df['Sector'])

(25,
 array(['Travel & Tourism', 'Consumer Services', nan,
        'Information Technology', 'Business Services', 'Insurance',
        'Finance', 'Retail', 'Media', 'Restaurants, Bars & Food Services',
        'Agriculture & Forestry', 'Non-Profit', 'Education', 'Government',
        'Health Care', 'Oil, Gas, Energy & Utilities',
        'Accounting & Legal', 'Manufacturing', 'Real Estate',
        'Biotech & Pharmaceuticals', 'Arts, Entertainment & Recreation',
        'Aerospace & Defense', 'Construction, Repair & Maintenance',
        'Transportation & Logistics', 'Telecommunications',
        'Mining & Metals'], dtype=object))

In [82]:
df.loc[~df.Industry.str.strip().duplicated(),'Industry'].count(), pd.unique(df['Industry'])

(126,
 array(['Travel Agencies', 'Health, Beauty, & Fitness', nan, 'Internet',
        'Advertising & Marketing', 'Insurance Carriers',
        'Investment Banking & Asset Management',
        'Enterprise Software & Network Solutions', 'Other Retail Stores',
        'Consulting', 'Research & Development',
        'Motion Picture Production & Distribution', 'Brokerage Services',
        'IT Services', 'Computer Hardware & Software', 'Security Services',
        'Casual Restaurants', 'Food & Beverage Stores', 'Food Production',
        'Health Fundraising Organizations',
        'Home Furniture & Housewares Stores',
        'Financial Analytics & Research', 'Education Training Services',
        'Lending', 'Social Assistance', 'Publishing', 'Federal Agencies',
        'Health Care Services & Hospitals',
        'Oil & Gas Exploration & Production', 'Staffing & Outsourcing',
        'Venture Capital & Private Equity',
        'Building & Personnel Services', 'Video Games', 'Accounting',
 

We have 25 unique values in `Sector` and 126 in `Industry`. They seem to describe the same thing, `Sector` being a more general and broader variable.

In [83]:
df['Type of ownership'].value_counts().count(), pd.unique(df['Type of ownership'])

(13,
 array(['Company - Private', nan, 'Company - Public',
        'Nonprofit Organization', 'Government', 'Hospital',
        'Self-employed', 'Subsidiary or Business Segment', 'Contract',
        'School / School District', 'Private Practice / Firm',
        'College / University', 'Franchise', 'Other Organization'],
       dtype=object))

In [84]:
df['Size'].value_counts().count(), pd.unique(df['Size'])

(7,
 array(['501 to 1000 employees', '1001 to 5000 employees',
        '1 to 50 employees', '201 to 500 employees', '51 to 200 employees',
        '10000+ employees', '5001 to 10000 employees', nan], dtype=object))

In [85]:
df['Revenue'].value_counts().count(), pd.unique(df['Revenue'])

(12,
 array([nan, '$100 to $500 million (USD)',
        '$500 million to $1 billion (USD)', '$10+ billion (USD)',
        '$25 to $50 million (USD)', '$1 to $5 million (USD)',
        '$5 to $10 billion (USD)', '$5 to $10 million (USD)',
        '$50 to $100 million (USD)', '$1 to $2 billion (USD)',
        '$2 to $5 billion (USD)', 'Less than $1 million (USD)',
        '$10 to $25 million (USD)'], dtype=object))

13 unique values for `Type of ownership`, 7 for `Size` and 12 for `Revenue'

Also, let's check `State` and `City`:

In [86]:
df.State.unique()

array(['NY', 'NJ', 'CA', 'IL', 'TX', 'AZ', 'DE', 'PA', 'United Kingdom',
       'FL', 'OH', 'UT', 'VA', 'NC', 'SC', 'IN', 'WA', 'GA', 'KS', 'CO',
       'Arapahoe', 'Los Angeles', 'TN', 'MD', 'DC', 'KY'], dtype=object)

In [87]:
df.City.unique()

array(['New York', 'Lyndhurst', 'Brooklyn', 'Jersey City', 'Carle Place',
       'Newark', 'Franklin Lakes', 'Port Washington', 'Rockville Centre',
       'Queens Village', 'Fort Lee', 'Middletown', 'Florham Park',
       'Summit', 'Maywood', 'West Orange', 'Los Angeles',
       'Woodland Hills', 'Universal City', 'Anaheim', 'Culver City',
       'Cerritos', 'Santa Monica', 'Torrance', 'El Segundo', 'Glendale',
       'Brea', 'Burbank', 'Westmont', 'Pasadena', 'Irwindale',
       'Calabasas', 'Hermosa Beach', 'Whittier', 'Northridge',
       'Marina del Rey', 'Monaco', 'Venice', 'Sherman Oaks', 'Duarte',
       'Wilmington', 'Long Beach', 'Monrovia', 'Carson', 'Seal Beach',
       'Monterey Park', 'Chicago', 'Evanston', 'Melrose Park', 'Lemont',
       'Rosemont', 'Northbrook', 'Elmhurst', 'Downers Grove',
       'Rolling Meadows', 'Naperville', 'Oak Brook', 'Des Plaines',
       'Hines', 'Evergreen Park', 'Bridgeview', 'Houston', 'Alvin',
       'Stafford', 'Pearland', 'Spring', 'Suga

We have few values that we need to fix in `State` (United Kindom: drop, Arapahoe: CO, Los Angeles: LA)

In [88]:
df.State.replace({'United Kingdom': np.NaN, 'Los Angeles': 'LA', 'Arapahoe': 'CO'}, inplace=True)
df.dropna(subset = ['State'],inplace=True)
df.State.unique()

array(['NY', 'NJ', 'CA', 'IL', 'TX', 'AZ', 'DE', 'PA', 'FL', 'OH', 'UT',
       'VA', 'NC', 'SC', 'IN', 'WA', 'GA', 'KS', 'CO', 'LA', 'TN', 'MD',
       'DC', 'KY'], dtype=object)

Let's make sure all the dtypes are correct:

In [89]:
df['State'] = df['State'].astype('category')
df['City'] = df['City'].astype('category')
df['In_Headquarters'] = df['In_Headquarters'].astype('category')
df['Employer_Estimate'] = df['Employer_Estimate'].astype('category')
df['Size'] = df['Size'].astype('category')
df['Sector'] = df['Sector'].astype('category')
df['Industry'] = df['Industry'].astype('category')
df['Type of ownership'] = df['Type of ownership'].astype('category')
df['Revenue'] = df['Revenue'].astype('category')

Now, let's start imputing!

First, we impute the easy ones. We created two categorical variables: `In_Headquarters` and `Employer_Estimate`, let's assume missing values as zeros:

In [90]:
df['In_Headquarters'].fillna(value=0, inplace=True)
df['Employer_Estimate'].fillna(value=0, inplace=True)

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11440 entries, 0 to 13155
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   Job_title          11440 non-null  object  
 1   Salary Estimate    9765 non-null   object  
 2   Job_Desc           11440 non-null  object  
 3   Rating             10391 non-null  float64 
 4   Company            11439 non-null  object  
 5   Location           9765 non-null   object  
 6   Headquarters       9161 non-null   object  
 7   Size               8990 non-null   category
 8   Founded            7202 non-null   object  
 9   Type of ownership  9096 non-null   category
 10  Industry           9977 non-null   category
 11  Sector             8413 non-null   category
 12  Revenue            6567 non-null   category
 13  In_Headquarters    11440 non-null  category
 14  City               11440 non-null  category
 15  State              11440 non-null  category
 16  Min_

In [92]:
100*df.Revenue.isnull().sum()/len(df)

42.59615384615385

`Revenue` is missing 42.5% of the data... We could still impute it. But before we impute, we need to set our criteria: What is going to be the determining factor to impute categorical variables?

First, let's create or final response variable `Avg_Salary`:

In [93]:
df['Avg_Salary'] = (df.Max_Salary + df.Min_Salary)/2

Now, we want our imputed observations to be good estimation of the real variable, that is: if our `Size` is 10000+, our salary might tend to be higher as bigger companies tend to pay higher salaries. Also, a Data Analyst will usually make less than a Senior Data Scientist, but maybe a Data Analyst working for a company with 100,000+ employees and a `Revenue` of 10+ billion (USD) will earn a similar amount to a Senior Data Scientist in a company with 50 employees and less than 1 million (USD) in revenue... So in order to impute these variables, we need to define our categories as dummy variables using `Job_title` and `Job_Desc`.

We want to categorize `Job_title` by: Position: Analyst, Engineer, Scientist, Deep (Learning), Machine (Learning). Senority: Manager, Senior or Sr., Junior or Jr., Lead, Principal, Associate, Chief, Intern.

In [94]:
df['Title_Analyst'] = df.apply(lambda x: 1 if 'analyst' in x.Job_title.lower() else 0, axis=1)
df['Title_Engineer'] = df.apply(lambda x: 1 if 'engineer' in x.Job_title.lower() else 0, axis=1)
df['Title_Scientist'] = df.apply(lambda x: 1 if 'scientist' in x.Job_title.lower() else 0, axis=1)
df['Title_Deep'] = df.apply(lambda x: 1 if 'deep' in x.Job_title.lower() else 0, axis=1)
df['Title_Machine'] = df.apply(lambda x: 1 if 'machine' in x.Job_title.lower() else 0, axis=1)
df['Title_Manager'] = df.apply(lambda x: 1 if 'manager' in x.Job_title.lower() else 0, axis=1)
df['Title_Senior'] = df.apply(lambda x: 1 if 'senior' in x.Job_title.lower() or 'sr' in x.Job_title.lower() else 0, axis=1)
df['Title_Junior'] = df.apply(lambda x: 1 if 'junior' in x.Job_title.lower() or 'jr' in x.Job_title.lower() else 0, axis=1)
df['Title_Lead'] = df.apply(lambda x: 1 if 'lead' in x.Job_title.lower() else 0, axis=1)
df['Title_Principal'] = df.apply(lambda x: 1 if 'principal' in x.Job_title.lower() else 0, axis=1)
df['Title_Chief'] = df.apply(lambda x: 1 if 'chief' in x.Job_title.lower() else 0, axis=1)
df['Title_Intern'] = df.apply(lambda x: 1 if 'intern' in x.Job_title.lower() else 0, axis=1)

In [95]:
df[['Title_Machine', 'Title_Deep', 'Title_Analyst','Title_Engineer','Title_Scientist','Title_Manager',
    'Title_Senior','Title_Junior','Title_Lead','Title_Principal','Title_Chief','Title_Intern']].value_counts()

Title_Machine  Title_Deep  Title_Analyst  Title_Engineer  Title_Scientist  Title_Manager  Title_Senior  Title_Junior  Title_Lead  Title_Principal  Title_Chief  Title_Intern
0              0           1              0               0                0              0             0             0           0                0            0               5378
                           0              0               1                0              0             0             0           0                0            0               1656
                           1              0               0                0              1             0             0           0                0            0                966
                           0              1               0                0              0             0             0           0                0            0                801
                                          0               1                0              1            

And now the same for `Job_Desc`. According to https://www.upgrad.com/blog/data-science-programming-languages/, these are the top 6 programming languages for Data Scientists: R, Python, SQL, Julia, Scala & JavaScript.

In [96]:
df['R'] = df.apply(lambda x: 1 if 'R.' in x.Job_Desc or 'R,' in x.Job_Desc else 0, axis=1)
df['Python'] = df.apply(lambda x: 1 if 'python' in x.Job_Desc.lower() else 0, axis=1)
df['Julia'] = df.apply(lambda x: 1 if 'julia' in x.Job_Desc.lower() else 0, axis=1)
df['SQL'] = df.apply(lambda x: 1 if 'sql' in x.Job_Desc.lower() else 0, axis=1)
df['Scala'] = df.apply(lambda x: 1 if 'scala' in x.Job_Desc.lower() else 0, axis=1)
df['JavaScript'] = df.apply(lambda x: 1 if 'javascript' in x.Job_Desc.lower() else 0, axis=1)

In [97]:
df[['R','Python','Julia','SQL','Scala','JavaScript']].value_counts()

R  Python  Julia  SQL  Scala  JavaScript
0  0       0      0    0      0             4415
                  1    0      0             1918
   1       0      1    0      0             1127
1  1       0      1    0      0              622
0  1       0      0    0      0              552
                  1    1      0              441
   0       0      0    1      0              386
1  0       0      0    0      0              307
   1       0      1    1      0              286
0  0       0      1    1      0              237
1  1       0      0    0      0              212
0  1       0      0    1      0              193
   0       0      1    0      1              166
1  0       0      1    0      0              149
0  1       0      1    0      1               66
1  1       0      0    1      0               64
0  1       0      1    1      1               44
1  1       0      1    0      1               39
0  1       0      0    0      1               35
1  0       0      0    1    

In [98]:
df = df.astype({'R':'category','Python':'category','Julia':'category','SQL':'category','Scala':'category',
                'JavaScript':'category','Title_Machine':'category', 'Title_Deep':'category', 'Title_Analyst':'category',
                'Title_Engineer':'category','Title_Scientist':'category','Title_Manager':'category',
    'Title_Senior':'category','Title_Junior':'category','Title_Lead':'category','Title_Principal':'category','Title_Chief':'category','Title_Intern':'category'})

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11440 entries, 0 to 13155
Data columns (total 38 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   Job_title          11440 non-null  object  
 1   Salary Estimate    9765 non-null   object  
 2   Job_Desc           11440 non-null  object  
 3   Rating             10391 non-null  float64 
 4   Company            11439 non-null  object  
 5   Location           9765 non-null   object  
 6   Headquarters       9161 non-null   object  
 7   Size               8990 non-null   category
 8   Founded            7202 non-null   object  
 9   Type of ownership  9096 non-null   category
 10  Industry           9977 non-null   category
 11  Sector             8413 non-null   category
 12  Revenue            6567 non-null   category
 13  In_Headquarters    11440 non-null  category
 14  City               11440 non-null  category
 15  State              11440 non-null  category
 16  Min_

Let's create or ordinal variables from `Size` and `Revenue`

In [100]:
df['Size_ord'] = df.Size.apply(lambda x: 0 if x == '1 to 50 employees' else(1 if x == '51 to 200 employees' else(2 if x == '201 to 500 employees'
                           else(3 if x == '501 to 1000 employees' else(4 if x == '1001 to 5000 employees' else(4 if x == '5001 to 10000 employees' else(5 if x == '10000+ employees' else np.NaN)))))))
df['Size_ord'] = df.Size_ord.astype('category')

df['Revenue_ord'] = df.Revenue.apply(lambda x: 0 if x == 'Less than $1 million (USD)' else(1 if x == '$1 to $5 million (USD)' else(2 if x == '$5 to $10 million (USD)'
                           else(3 if x == '$10 to $25 million (USD)' else(4 if x == '$25 to $50 million (USD)' else(5 if x == '$50 to $100 million (USD)' 
                           else(6 if x == '$100 to $500 million (USD)' else(7 if x == '$500 million to $1 billion (USD)' else(8 if x == '$1 to $2 billion (USD)' else(9 if x == '$2 to $5 billion (USD)'
                           else(10 if x == '$5 to $10 billion (USD)' else(11 if x == '$10+ billion (USD)' else np.NaN))))))))))))
df['Revenue_ord'] = df.Revenue_ord.astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11440 entries, 0 to 13155
Data columns (total 40 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   Job_title          11440 non-null  object  
 1   Salary Estimate    9765 non-null   object  
 2   Job_Desc           11440 non-null  object  
 3   Rating             10391 non-null  float64 
 4   Company            11439 non-null  object  
 5   Location           9765 non-null   object  
 6   Headquarters       9161 non-null   object  
 7   Size               8990 non-null   category
 8   Founded            7202 non-null   object  
 9   Type of ownership  9096 non-null   category
 10  Industry           9977 non-null   category
 11  Sector             8413 non-null   category
 12  Revenue            6567 non-null   category
 13  In_Headquarters    11440 non-null  category
 14  City               11440 non-null  category
 15  State              11440 non-null  category
 16  Min_

Now we are ready to impute:

In [101]:
from sklearn.impute import KNNImputer
imputer = KNNImputer()
df_imputed = pd.DataFrame(np.round(imputer.fit_transform(df[['R','Python','Julia','SQL','Scala','JavaScript','Title_Machine', 'Title_Deep', 'Title_Analyst','Title_Engineer','Title_Scientist','Title_Manager',
    'Title_Senior','Title_Junior','Title_Lead','Title_Principal','Title_Chief','Title_Intern', 'Avg_Salary','Employer_Estimate','Size_ord', 'Revenue_ord']])), columns = ['R','Python','Julia','SQL','Scala','JavaScript','Title_Machine', 'Title_Deep', 'Title_Analyst','Title_Engineer','Title_Scientist','Title_Manager',
    'Title_Senior','Title_Junior','Title_Lead','Title_Principal','Title_Chief','Title_Intern', 'Avg_Salary', 'Employer_Estimate', 'Size_ord', 'Revenue_ord'])


In [102]:
df_imputed

R  Python  Julia  SQL  Scala  JavaScript  Title_Machine  Title_Deep  \
0      1.0     0.0    0.0  1.0    0.0         0.0            0.0         0.0   
1      0.0     1.0    0.0  1.0    0.0         0.0            0.0         0.0   
2      1.0     1.0    0.0  0.0    0.0         0.0            0.0         0.0   
3      0.0     1.0    0.0  1.0    0.0         0.0            0.0         0.0   
4      1.0     1.0    0.0  0.0    0.0         0.0            0.0         0.0   
...    ...     ...    ...  ...    ...         ...            ...         ...   
11435  0.0     0.0    0.0  0.0    0.0         0.0            0.0         0.0   
11436  1.0     1.0    0.0  0.0    0.0         0.0            0.0         1.0   
11437  0.0     0.0    0.0  0.0    0.0         0.0            0.0         0.0   
11438  0.0     0.0    0.0  0.0    0.0         0.0            0.0         0.0   
11439  0.0     0.0    0.0  1.0    1.0         0.0            0.0         0.0   

       Title_Analyst  Title_Engineer  ...  Title_Senior  Title_Junior  \
0                0.0             0.0  ...           1.0           0.0   
1                0.0             0.0  ...           0.0           0.0   
2                0.0             0.0  ...           0.0           0.0   
3                1.0             0.0  ...           0.0           0.0   
4                0.0             0.0  ...           0.0           0.0   
...              ...             ...  ...           ...           ...   
11435            1.0             0.0  ...           0.0           0.0   
11436            0.0             0.0  ...           0.0           0.0   
11437            0.0             0.0  ...           0.0           0.0   
11438            0.0             0.0  ...           0.0           0.0   
11439            0.0             1.0  ...           1.0           0.0   

       Title_Lead  Title_Principal  Title_Chief  Title_Intern  Avg_Salary  \
0             0.0              0.0          0.0           0.0    146000.0   
1             0.0              0.0          0.0           0.0    146000.0   
2             0.0              0.0          0.0           0.0    146000.0   
3             0.0              0.0          0.0           0.0    146000.0   
4             0.0              0.0          0.0           0.0    146000.0   
...           ...              ...          ...           ...         ...   
11435         0.0              0.0          0.0           0.0     73161.0   
11436         0.0              0.0          0.0           0.0     99885.0   
11437         0.0              0.0          0.0           0.0    129154.0   
11438         0.0              0.0          0.0           0.0    143411.0   
11439         0.0              0.0          0.0           0.0    135675.0   

       Employer_Estimate  Size_ord  Revenue_ord  
0                    0.0       3.0          7.0  
1                    0.0       4.0          6.0  
2                    0.0       0.0          5.0  
3                    0.0       2.0          5.0  
4                    0.0       1.0          5.0  
...                  ...       ...          ...  
11435                0.0       2.0          7.0  
11436                0.0       3.0          4.0  
11437                0.0       4.0          7.0  
11438                0.0       3.0          8.0  
11439                0.0       3.0          7.0  

[11440 rows x 22 columns]

In [103]:
df.reset_index(inplace=True)
sector_dummy = pd.get_dummies(df.Sector)
sector_dummy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11440 entries, 0 to 11439
Data columns (total 25 columns):
 #   Column                              Non-Null Count  Dtype
---  ------                              --------------  -----
 0   Accounting & Legal                  11440 non-null  uint8
 1   Aerospace & Defense                 11440 non-null  uint8
 2   Agriculture & Forestry              11440 non-null  uint8
 3   Arts, Entertainment & Recreation    11440 non-null  uint8
 4   Biotech & Pharmaceuticals           11440 non-null  uint8
 5   Business Services                   11440 non-null  uint8
 6   Construction, Repair & Maintenance  11440 non-null  uint8
 7   Consumer Services                   11440 non-null  uint8
 8   Education                           11440 non-null  uint8
 9   Finance                             11440 non-null  uint8
 10  Government                          11440 non-null  uint8
 11  Health Care                         11440 non-null  uint8
 12  Info

In [104]:
sector_dummy

Accounting & Legal  Aerospace & Defense  Agriculture & Forestry  \
0                       0                    0                       0   
1                       0                    0                       0   
2                       0                    0                       0   
3                       0                    0                       0   
4                       0                    0                       0   
...                   ...                  ...                     ...   
11435                   0                    0                       0   
11436                   0                    0                       0   
11437                   0                    0                       0   
11438                   0                    0                       0   
11439                   0                    0                       0   

       Arts, Entertainment & Recreation  Biotech & Pharmaceuticals  \
0                                     0                          0   
1                                     0                          0   
2                                     0                          0   
3                                     0                          0   
4                                     0                          0   
...                                 ...                        ...   
11435                                 0                          0   
11436                                 0                          0   
11437                                 0                          0   
11438                                 0                          0   
11439                                 0                          0   

       Business Services  Construction, Repair & Maintenance  \
0                      0                                   0   
1                      0                                   0   
2                      0                                   0   
3                      0                                   0   
4                      1                                   0   
...                  ...                                 ...   
11435                  0                                   0   
11436                  0                                   0   
11437                  0                                   0   
11438                  0                                   0   
11439                  0                                   0   

       Consumer Services  Education  Finance  ...  Media  Mining & Metals  \
0                      0          0        0  ...      0                0   
1                      1          0        0  ...      0                0   
2                      0          0        0  ...      0                0   
3                      0          0        0  ...      0                0   
4                      0          0        0  ...      0                0   
...                  ...        ...      ...  ...    ...              ...   
11435                  0          0        0  ...      0                0   
11436                  0          0        0  ...      0                0   
11437                  0          0        0  ...      0                0   
11438                  0          0        0  ...      0                0   
11439                  0          0        0  ...      0                0   

       Non-Profit  Oil, Gas, Energy & Utilities  Real Estate  \
0               0                             0            0   
1               0                             0            0   
2               0                             0            0   
3               0                             0            0   
4               0                             0            0   
...           ...                           ...          ...   
11435           0                             0            0   
11436           0                             0            0   
11437  

Need dummies for `State`, `City`. 

In [105]:
state_dummy = pd.get_dummies(df.State)
state_dummy

AZ  CA  CO  DC  DE  FL  GA  IL  IN  KS  ...  NJ  NY  OH  PA  SC  TN  \
0       0   0   0   0   0   0   0   0   0   0  ...   0   1   0   0   0   0   
1       0   0   0   0   0   0   0   0   0   0  ...   0   1   0   0   0   0   
2       0   0   0   0   0   0   0   0   0   0  ...   0   1   0   0   0   0   
3       0   0   0   0   0   0   0   0   0   0  ...   1   0   0   0   0   0   
4       0   0   0   0   0   0   0   0   0   0  ...   0   1   0   0   0   0   
...    ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..   
11435   0   1   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
11436   0   1   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
11437   0   1   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
11438   0   1   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
11439   0   1   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   

       TX  UT  VA  WA  
0       0   0   0   0  
1       0   0   0   0  
2       0   0   0   0  
3       0   0   0   0  
4       0   0   0   0  
...    ..  ..  ..  ..  
11435   0   0   0   0  
11436   0   0   0   0  
11437   0   0   0   0  
11438   0   0   0   0  
11439   0   0   0   0  

[11440 rows x 24 columns]

In [106]:
city_dummy = pd.get_dummies(df.City)
city_dummy.shape

(11440, 437)

437 dummy variables... maybe not.

In [107]:
del city_dummy

Let's put it all together:

In [111]:
df_out = pd.concat([df.Job_title,df_imputed, state_dummy, sector_dummy],axis=1)

Let's save our data:

In [112]:
df_out.to_csv('Data_Wrangled.csv')